### PR TITLE
Make action compatible with private repos.

### PR DESCRIPTION
**Pull request by ApoorvGuptaAi**
# Make action compatible with private repos.

 - Switches to the latest github actions package and API.
 - Use octokit rest API interface to get PR diff (works for private repos too).

[x] Unit tests run and pass.

### CODE REVIEW

**Comment by ApoorvGuptaAi** [comment]
Example run on private repo:

Screenshot 2023-11-06 at 9 18 02 PM

**Comment by JJ** [comment]
> # Make action compatible with private repos.
> 
>     * Switches to the latest github actions package and API.
> 
>     * Use octokit rest API interface to get PR diff (works for private repos too).

Wow, amazing. Thanks a lot. I need some time to check it out, even to understand it properly. Stick around, don't go far, and again, thanks a lot.

File: src/main.ts [hunk 1]
```diff
@@ -1,21 +1,44 @@
 import * as core from "@actions/core";
-const { GitHub, context } = require("@actions/github");
+import { getOctokit, context } from "@actions/github";
 import parse from "parse-diff";
 import { rexify } from "./utils";
 
+async function getDiff(octokit, context) {
```
Comment (JJ) [inline src/main.ts:6]: What we are using here from context is payload.repository and payload.pull request. Maybe destructure  these in the argument declaration?

**Comment by ApoorvGuptaAi** [comment]
> Wow, amazing. Thanks a lot. I need some time to check it out, even to understand it properly. Stick around, don't go far, and again, thanks a lot.

Couple of things to look out for:

- Havent tested the waivedUser functionality, specifically if non-"User" type users should be checked.
- Tested on small PRs on private repos, I dont know if the new diff API has any limitations on size.

File: src/main.ts [hunk 2]
```diff
@@ -1,21 +1,44 @@
...
-    // First check for waived users
-    const waivedUsers = core.getInput("waivedUsers") || ["dependabot[bot]"];
+    const octokit = getOctokit(token);
 
-    if (waivedUsers.includes(user)) {
-      core.warning(`⚠️ Not running this workflow for waived user «${user}»`);
-      return;
-    }
+    const senderInfo = context?.payload?.sender;
+    const senderName = senderInfo?.login;
```
Comment (JJ) [inline src/main.ts:29]: If senderInfo is null, maybe this should bail out right here, it's going to be impossible to follow...

File: src/main.ts [hunk 3]
```diff
@@ -1,21 +1,44 @@
...
+    const senderType = senderInfo?.type
+    core.info(`PR created by ${senderName} (${senderType})`)
```
Comment (JJ) [inline src/main.ts:31]: First time I find this senderType thing. I'll check it out.

File: src/main.ts [hunk 4]
```diff
@@ -1,21 +1,44 @@
...
+    // NOTE(ApoorvGuptaAI): Maybe we should not check senderType before doing the waive check.
+    if (senderType === 'User') {
+      // First check for waived users
+      const waivedUsers = core.getInput("waivedUsers") || ["dependabot[bot]"];
+
+      if (waivedUsers.includes(senderName)) {
+        core.warning(`⚠️ Not running this workflow for waived user «${senderName}»`);
+        return;
+      }
+  }
```
Comment (JJ) [inline src/main.ts:41]: No else here? What happens if it's not an user? Maybe warn "waivedUsers" is not applicable?

File: src/main.ts [hunk 5]
```diff
@@ -28,13 +51,11 @@ async function run() {
       core.warning("⚠️ Not a pull request, skipping PR body checks");
     } else {
       if (bodyContains || bodyDoesNotContain) {
+        const PRBody = context?.payload?.pull_request?.body;
```
Comment (JJ) [inline src/main.ts:54]: Again, the action can't continue if context does not exist; if payload does not exist, either. It's probably best to check early, and then just go with it, don't you think so?

File: src/main.ts [hunk 6]
```diff
@@ -48,64 +69,56 @@ async function run() {
...
-        if (filesChanged && files.length != filesChanged) {
-          core.setFailed(
-            "You should change exactly " + filesChanged + " file(s)"
-          );
-        }
+      core.info("Checking diff contents");
+      const diffContains = core.getInput("diffContains");
+      const diffDoesNotContain = core.getInput("diffDoesNotContain");
+
+      const files = await getDiff(octokit, context);
```
Comment (JJ) [inline src/main.ts:76]: So what you've done here is factor out diff processing to the function. Good one.

File: src/main.ts [hunk 7]
```diff
@@ -48,64 +69,56 @@ async function run() {
...
-            chunk.changes.forEach(function (change: any) {
-              if (change.add) {
-                changes += change.content;
-              }
-            });
+      let changes = "";
+      let additions: number = 0;
+      files.forEach(function (file) {
+        additions += file.additions;
+        file.chunks.forEach(function (chunk: parse.Chunk) {
```
Comment (JJ) [inline src/main.ts:90]: No change here but indentation.

File: src/main.ts [hunk 8]
```diff
@@ -48,64 +69,56 @@ async function run() {
...
+        core.exportVariable("diff", changes);
+        core.setOutput("diff", changes);
+      }
+      if (diffDoesNotContain && rexify(diffDoesNotContain).test(changes)) {
+        core.setFailed(
+          "The added code should not contain " + diffDoesNotContain
         );
       }
+
+      core.info("Checking lines/files changed");
```
Comment (JJ) [inline src/main.ts:112]: Don't know what happened here, seems to be the same except for indentation...

**Review by JJ — APPROVED** [review]
I would say it looks fine overall. There are some minor nits, would be helpful if you clarified them before merging.

**Comment by JJ** [comment]
> > Wow, amazing. Thanks a lot. I need some time to check it out, even to understand it properly. Stick around, don't go far, and again, thanks a lot.
> 
> Couple of things to look out for:
> 
>     * Havent tested the waivedUser functionality, specifically if non-"User" type users should be checked.

That was one of my questions, probably something should be conveyed to the user in that case, but so far I didn't know such kind of users existed... Might be worth the while a separate PR, maybe, although what is there so far looks great.
> 
>     * Tested on small PRs on private repos, I dont know if the new diff API has any limitations on size

In diff size, you mean? At any rate, maybe add a caveat in the documentation?

**Comment by JJ** [comment]
Also, waivedUsers is not working. I have to get around to it since summer #106

**Comment by ApoorvGuptaAi** [comment]
@JJ Havent tested the latest set of changes, is there some way to test it out?

**Comment by JJ** [comment]
Just the basics, I just approved running them. I don't have time ATM but I would say all looks good... Thanks!

**Comment by ApoorvGuptaAi** [comment]
Just confirmed itsworking, please go ahead and merge.

**Review by JJ — APPROVED** [review]
Still a `TODO` and `NOTE` to go, but I guess your intention is to create an issue for later, right?

**Pull request merged**